virtual_rainforest/core/registry.py

```diff
@@ -0,0 +1,222 @@
+"""The :mod:`~virtual_rainforest.core.registry` module is used to populate the
+:data:`~virtual_rainforest.core.registry.MODULE_REGISTRY`. This provides a dictionary
+for the core components of Virtual Rainforest modules that is used in model
+configuration. The core components included in the registry are:
+
+* The module schema, used to validate configuration data for running a simulation.
+* Any module constants classes, which are then available for automatic validation
+    and instance creation.
+* The model class associated with the module.
+
+Note that the  :mod:`virtual_rainforest.core` module does not have an associated model
+but does have a schema and constants.
+
+The module provides the :func:`~virtual_rainforest.core.registry.register_module`
+umbrella function and then separate functions to handle the components:
+:func:`~virtual_rainforest.core.registry.register_schema`,
+:func:`~virtual_rainforest.core.registry.register_model` and
+:func:`~virtual_rainforest.core.registry.register_constants_classes`.
+"""  # noqa: D205, D415
+
+
+from dataclasses import is_dataclass
+from importlib import import_module, resources
+from inspect import getmembers, isclass
+from pathlib import Path
+from types import ModuleType
+from typing import Any
+
+from virtual_rainforest.core.logger import LOGGER
+from virtual_rainforest.core.schema import load_schema
+
+MODULE_REGISTRY: dict[str, Any] = {}
+"""Information about module schemas, constants and model, keyed by module name.
+
+The registry is a dictionary keyed by the final part of the module name. The entry for
+each module is itself a dictionary with 'model', 'constants' and 'schema' entries. For
+example:
+
+MODULE_REGISTRY['plants']['model'] = <virtual_rainforest.models.plants.PlantsModel>
+MODULE_REGISTRY['plants']['constants'] = [
+    'PlantsConsts': <virtual_rainforest.models.plants.constants.PlantsConsts>
+]
+MODULE_REGISTRY['plants']['schema'] = {schema_dict}
+"""
```

Comment:
If each element of `MODEL_REGISTRY` is a dict with the same members, couldn't you use a dataclass for this instead? That way you get better typing etc.

Reply:
Meant to raise this point - you had this structure in an earlier post. The problem I ran into here is that importing those types to construct the dataclass causes some horrific circularity issues. Part of the reason I have separated `schema` and `registry` here is that they are then very low level and don't require much else from VR. Once you import `BaseModel` or things like it then I was getting circular imports all over the place.

It would certainly be much neater to type this more explicitly.

Reply:
Couldn't the relevant types be put into their own `.py` files then? I can't imagine the `MODEL_REGISTRY` would be needed in those files, so there shouldn't be any circular import problems.

Reply:
The issue is that `BaseModel` has a whole bunch of higher level imports:

```
from virtual_rainforest.core.axes import AXIS_VALIDATORS
from virtual_rainforest.core.config import Config
from virtual_rainforest.core.data import Data
```

All of those structures come after configuration, so the circular imports from importing `BaseModel` seem really hard to avoid.